src/base/initialize.js

```diff
@@ -0,0 +1,89 @@
+import { createBrowserHistory } from 'history';
+import {
+  configure as configurePubSub,
+  publish,
+  getPubSubService,
+  PubSubJsService,
+} from '../pubSub';
+import { configure as configureConfig, ProcessEnvConfigService, getConfigService } from '../config';
+import { configureLogging, getLoggingService, NewRelicLoggingService } from '../logging';
+import { auth, initError } from './handlers';
+
+export const APP_TOPIC = 'APP';
+export const APP_PUBSUB_INITIALIZED = `${APP_TOPIC}.PUBSUB_INITIALIZED`;
+export const APP_CONFIG_INITIALIZED = `${APP_TOPIC}.CONFIG_INITIALIZED`;
+export const APP_AUTH_INITIALIZED = `${APP_TOPIC}.AUTH_INITIALIZED`;
+export const APP_I18N_INITIALIZED = `${APP_TOPIC}.I18N_INITIALIZED`;
+export const APP_LOGGING_INITIALIZED = `${APP_TOPIC}.LOGGING_INITIALIZED`;
+export const APP_ANALYTICS_INITIALIZED = `${APP_TOPIC}.ANALYTICS_INITIALIZED`;
+export const APP_BEFORE_READY = `${APP_TOPIC}.BEFORE_READY`;
+export const APP_READY = `${APP_TOPIC}.READY`;
+export const APP_INIT_ERROR = `${APP_TOPIC}.INIT_ERROR`;
+
+const override = async (defaultHandler, overrideHandler, ...args) => {
+  if (overrideHandler !== undefined) {
+    await overrideHandler(args);
+  } else {
+    await defaultHandler(args);
+  }
+};
+
+export const history = createBrowserHistory();
+
+export default async function initialize({
+  pubSubService = PubSubJsService,
+  configService = ProcessEnvConfigService,
+  loggingService = NewRelicLoggingService,
+  analyticsService = null,
+  authService = null,
+  i18nService = null,
+  requireAuthenticatedUser = false,
+  hydrateAuthenticatedUser = false,
+  messages,
+  handlers = {},
+} = {}) {
+  try {
+    configurePubSub(pubSubService);
+    await override(() => {}, handlers.pubSub);
```

Comment:
This override function is a little strange to me at first glance. What dropping it and using defaults instead – something like:
```
const handlers = Object.create(defaultHandlers, _handlers);

// ...
configureConfig(configService, { pubSubService: getPubSubService() });
handlers.config();
publish(APP_CONFIG_INITIALIZED);
```

Reply:
Cool - I like that.